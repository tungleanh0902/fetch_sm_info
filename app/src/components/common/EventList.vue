<template>
  <div v-infinite-scroll="loadMore" infinite-scroll-distance="80">
    <template v-for="group in groupedEvents">
      <div :key="group[0].section">
        <h3>{{ group[0].section }}</h3>
        <template v-for="item in group">
          <slot v-bind="item.event" />
        </template>
      </div>
    </template>
  </div>
</template>

<script>
import groupBy from "lodash.groupby"
import orderBy from "lodash.orderby"
import moment from "moment"

const categories = [
  {
    section: "Today",
    matcher: (event) => {
      // tests if the timestamp has the same day as today
      return moment(event.timestamp).isSame(moment(), "day")
    },
  },
  {
    section: "Yesterday",
    matcher: (event) => {
      return moment(event.timestamp).isSame(moment().subtract(1, "days"), "day")
    },
  },
]

export default {
  name: `event-list`,
  props: {
    events: {
      type: Array,
      required: true,
    },
    moreAvailable: {
      type: Boolean,
      required: true,
    },
  },
  data: () => ({
    showing: 15,
    maxReached: false,
  }),
  computed: {
    showingEvents() {
      return this.events.slice(0, this.showing)
    },
    groupedEvents() {
      return orderBy(
        groupBy(this.categorizedEvents, "section"),
        (group) => group[0].event.timestamp,
        "desc"
      )
    },
    categorizedEvents() {
      return this.showingEvents.map((event) => {
        // check if the tx is in Today, Yesterday or Last Week
        const dateString =
          ` (` + moment(event.timestamp).format("MMMM Do") + `)`
        const category = categories.find(({ matcher }) => matcher(event))
        if (category) {
          return {
            section: category.section + dateString,
            event,
          }
        }

        // check if tx is in a month this year
        const date = moment(event.timestamp)
        const today = moment()
        if (date.year() === today.year()) {
          return {
            section: date.format("MMMM Do"),
            event,
          }
        }

        // tx is in a month another year
        return {
          section: date.format("MMMM Do, YYYY"),
          event,
        }
      })
    },
  },
  methods: {
    loadMore() {
      if (!this.maxReached) {
        this.showing += 50

        if (this.showing > this.events.length - 100 && !this.moreAvailable) {
          this.maxReached = true
        }

        // preload next transactions before scroll end and check if last loading loads new records
        if (this.showing > this.events.length - 100 && this.moreAvailable) {
          this.$emit("loadMore")
        }
      }
    },
  },
}
</script>
<style scoped>
h3 {
  margin: 2rem 0 0.5rem 1.5rem;
  color: var(--dim);
  font-size: var(--sm);
  font-weight: 500;
}
</style>
