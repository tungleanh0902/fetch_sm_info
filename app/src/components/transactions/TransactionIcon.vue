<template>
  <div class="tx__icon">
    <img :src="icon" :alt="transactionType" />
  </div>
</template>

<script>
export default {
  name: `transaction-icon`,
  props: {
    transactionType: {
      type: String,
      required: true,
    },
  },
  computed: {
    icon() {
      return `/img/icons/activity/${this.transactionType}.svg`
    },
  },
}
</script>

<style>
.tx__icon {
  padding: 1rem;
}

.tx__icon img {
  max-height: 100%;
  max-width: 40px;
  display: block;
}

@media screen and (min-width: 1023px) {
  .tx__icon img {
    max-width: 48px;
  }
}
</style>
