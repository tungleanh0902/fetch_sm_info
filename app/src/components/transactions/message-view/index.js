import SendTxDetails from "./SendTxDetails"
import StakeTxDetails from "./StakeTxDetails"
import RestakeTxDetails from "./RestakeTxDetails"
import UnstakeTxDetails from "./UnstakeTxDetails"
import ClaimRewardsTxDetails from "./ClaimRewardsTxDetails"
import SubmitProposalTxDetails from "./SubmitProposalTxDetails"
import DepositTxDetails from "./DepositTxDetails"
import VoteTxDetails from "./VoteTxDetails"
import UnknownTxDetails from "./UnknownTxDetails"

export {
  SendTxDetails,
  StakeTxDetails,
  RestakeTxDetails,
  UnstakeTxDetails,
  ClaimRewardsTxDetails,
  SubmitProposalTxDetails,
  DepositTxDetails,
  VoteTxDetails,
  UnknownTxDetails,
}
