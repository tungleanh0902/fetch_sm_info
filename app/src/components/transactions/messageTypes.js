const messageType = {
  SEND: "SendTx",
  STAKE: "StakeTx",
  RESTAKE: "RestakeTx",
  UNSTAKE: "UnstakeTx",
  SUBMIT_PROPOSAL: "SubmitProposalTx",
  DEPOSIT: "DepositTx",
  VOTE: "VoteTx",
  CLAIM_REWARDS: "ClaimRewardsTx",
  WITHDRAW_UNSTAKED: "WithdrawUnstakedTokensTx",
  UNKNOWN: "UnknownTx",
}

export { messageType }
