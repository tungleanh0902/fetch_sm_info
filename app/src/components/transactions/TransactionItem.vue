<template>
  <div class="tx-container">
    <div class="tx" @click="toggleDetail">
      <component
        :is="messageTypeComponent"
        :transaction="transaction"
        :validators="validators"
        :session-address="address"
        :show="show"
      />
      <div v-if="!isExtension" class="toggle" :class="{ up: show }">
        <i class="material-icons notranslate toggle-icon"
          >keyboard_arrow_down</i
        >
      </div>
    </div>
    <transition v-if="!isExtension" name="slide-out">
      <div v-if="show" class="tx-details">
        <TransactionMetadata v-if="showMetaData" :transaction="transaction" />
      </div>
    </transition>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import { messageType } from "./messageTypes.js"
import TransactionMetadata from "./TransactionMetadata"

import {
  SendTxDetails,
  StakeTxDetails,
  RestakeTxDetails,
  UnstakeTxDetails,
  ClaimRewardsTxDetails,
  SubmitProposalTxDetails,
  DepositTxDetails,
  VoteTxDetails,
  UnknownTxDetails,
} from "./message-view"

export default {
  name: `tx-item`,
  components: {
    TransactionMetadata,
    SendTxDetails,
    StakeTxDetails,
    RestakeTxDetails,
    UnstakeTxDetails,
    ClaimRewardsTxDetails,
    SubmitProposalTxDetails,
    DepositTxDetails,
    VoteTxDetails,
    UnknownTxDetails,
  },
  props: {
    transaction: {
      type: Object,
      required: true,
    },
    validators: {
      type: Object,
      required: true,
    },
    address: {
      type: String,
      default: null,
    },
    showMetaData: {
      type: Boolean,
      default: true,
    },
  },
  data: () => ({
    show: false,
  }),
  computed: {
    ...mapGetters([`isExtension`]),
    messageTypeComponent: function () {
      switch (this.transaction.type) {
        case messageType.SEND:
          return `send-tx-details`
        case messageType.STAKE:
          return `stake-tx-details`
        case messageType.UNSTAKE:
          return `unstake-tx-details`
        case messageType.RESTAKE:
          return `restake-tx-details`
        case messageType.SUBMIT_PROPOSAL:
          return `submit-proposal-tx-details`
        case messageType.DEPOSIT:
          return `deposit-tx-details`
        case messageType.VOTE:
          return `vote-tx-details`
        case messageType.CLAIM_REWARDS:
          return `claim-rewards-tx-details`
        case messageType.UNKNOWN:
          return `unknown-tx-details`
        /* istanbul ignore next */
        default:
          return ``
      }
    },
  },
  methods: {
    toggleDetail(event) {
      if (event.target.className !== `address`) {
        this.show = !this.show
      }
    },
  },
}
</script>

<style>
.tx {
  position: relative;
  font-size: 14px;
  display: flex;
  align-items: center;
  background: var(--app-fg);
  border: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
  z-index: 90;
  cursor: pointer;
}

.tx a {
  display: inline-block;
}

.validator-image {
  margin: 0.25rem;
  border-radius: 100%;
  height: 1.25rem;
  width: 1.25rem;
  vertical-align: middle;
  transition: transform 0.2s ease-in-out;
}

.validator-image svg {
  border-radius: 100%;
}

.tx a:hover .validator-image {
  transform: scale(1.1);
}

.tx h3 {
  font-size: 14px;
  font-weight: 400;
  color: var(--bright);
}

.tx .amount {
  white-space: nowrap;
}

.tx-container {
  margin: 0 1rem 0.5rem;
}

.tx-details {
  background: var(--app-fg);
  border-left: 1px solid var(--bc-dim);
  border-right: 1px solid var(--bc-dim);
  border-bottom: 1px solid var(--bc-dim);
  border-bottom-left-radius: 0.25rem;
  border-bottom-right-radius: 0.25rem;
  margin: 0 1rem 0.5rem 1rem;
  font-size: 14px;
  padding: 1rem;
  position: relative;
  z-index: 0;
}

.tx__content {
  display: flex;
  align-items: center;
  position: relative;
  z-index: 90;
  width: 100%;
}

.tx__content__left {
  padding: 1rem 1rem 1rem 0;
}

.tx__content__right {
  padding-right: 2rem;
  text-align: right;
  flex: auto;
}

.slide-out-enter-active,
.slide-out-leave-active {
  transition: all 0.2s;
}

.slide-out-enter,
.slide-out-leave-to {
  opacity: 0;
  transform: translateY(-100px);
  margin-bottom: -100px;
}

.toggle {
  position: relative;
  right: 1rem;
  z-index: 91;
  cursor: pointer;
  padding: 0.1rem;
  border-radius: 50%;
  background: var(--bc-dim);
  height: 20px;
  width: 20px;
  transition: transform 0.2s ease;
}

.toggle.up {
  transform: rotate(180deg);
}

.toggle-icon {
  font-size: 16px;
}

.tx .copied {
  position: absolute;
  top: -0.5rem;
  right: 0;
}

@media screen and (max-width: 767px) {
  .tx__icon {
    display: none;
  }

  .tx__content__left {
    padding-left: 1rem;
  }

  .toggle {
    display: none;
  }

  .amount {
    position: absolute;
    right: 1rem;
    top: 1rem;
  }

  .tx-details {
    margin: 0 0.25rem 0.25rem;
  }

  .validator-image {
    height: 1rem;
    width: 1rem;
  }
}
</style>
