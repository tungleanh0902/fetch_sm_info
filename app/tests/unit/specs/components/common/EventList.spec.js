import { shallowMount } from "@vue/test-utils"
import EventList from "src/components/common/EventList"
import { messageType } from "src/components/transactions/messageTypes"

describe(`EventList`, () => {
  let wrapper

  let txs = []
  let height = 1000000

  for (var type in messageType) {
    txs.push({
      type: messageType[type],
      hash: "A0DEB29E97A4DF38289D55D63C5724588985E1D35B26518CB66EAF96CFEF2E04",
      height,
      timestamp: new Date(
        Date.now() + txs.length * 1000 * 60 * 60 * 24
      ).toISOString(),
      memo: "",
      success: true,
      fees: [],
      details: {
        from: ["cosmosvaloper123"],
        to: ["cosmosvaloper456"],
        proposalType: "cosmos-sdk/TextProposal",
        proposalTitle: "Lunie was here!",
        proposalDescription: "Blah blah blah and blah blah blah ...",
        proposalId: 17,
        voteOption: "Yes",
        initialDeposit: {
          denom: "ATOM",
          amount: "10",
        },
        amount: {
          denom: "ATOM",
          amount: "10",
        },
      },
    })
  }

  beforeEach(() => {
    wrapper = shallowMount(EventList, {
      propsData: {
        events: txs,
        moreAvailable: true,
      },
      directives: {
        infiniteScroll: () => jest.fn(),
      },
      slots: {
        default: `<div />`,
      },
    })
  })

  it(`calls loadMore script on scroll`, () => {
    const self = {
      $emit: jest.fn(),
      events: txs,
      moreAvailable: true,
      maxReached: false,
      showing: 1,
    }
    EventList.methods.loadMore.call(self)
    expect(self.$emit).toHaveBeenCalledWith(`loadMore`)
  })

  it(`doesn't call loadMore if not more available`, () => {
    const self = {
      $emit: jest.fn(),
      events: txs,
      moreAvailable: false,
      maxReached: false,
    }
    EventList.methods.loadMore.call(self)
    expect(self.$emit).not.toHaveBeenCalledWith(`loadMore`)
  })

  it(`doesn't call loadMore if already having enough`, () => {
    const self = {
      $emit: jest.fn(),
      moreAvailable: false,
      maxReached: false,
      events: new Array(20),
      showing: 20,
    }
    EventList.methods.loadMore.call(self)
    expect(self.$emit).not.toHaveBeenCalledWith(`loadMore`)
  })

  it(`renders a list of event items`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
