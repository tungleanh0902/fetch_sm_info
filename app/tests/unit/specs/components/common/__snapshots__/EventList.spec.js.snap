// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventList renders a list of event items 1`] = `
<div
  infinite-scroll-distance="80"
>
  <div>
    <h3>
      January 10th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 9th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 8th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 7th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 6th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 5th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 4th
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 3rd
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      January 2nd
    </h3>
     
    <div />
  </div>
  <div>
    <h3>
      Today (January 1st)
    </h3>
     
    <div />
  </div>
</div>
`;
